### PR TITLE
Compute trace Christoffel from GH vars better

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add function that is optimized (doesn't do excessive FLOPs at least)
- Use it when computing the 1-index constraint for observing

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp [hunk 1]
```diff
@@ -44,6 +44,81 @@ auto christoffel_second_kind(
...
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) {
+  destructive_resize_components(trace,
+                                get_size(get<0>(spacetime_normal_one_form)));
+  get<0>(*trace) = 0.0;
+  // Compute common terms between components.
+  for (size_t b = 0; b < SpatialDim + 1; ++b) {
```
Comment (knelli2) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp:76]: If we want to start using tensor expressions in more places, maybe consider changing this to use tensor expressions? Up to you though!

**Review by knelli2 — DISMISSED** [review]

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp (hunk 1 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp:76]: Yea, I thought about that but want to just keep this as-is for now. We need to do a full performance audit on GH anyway and should explore TEs then

Comment (knelli2) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp:76]: Ok, works for me!

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.hpp [hunk 2]
```diff
@@ -40,4 +40,38 @@ auto christoffel_second_kind(
     const tnsr::II<DataType, SpatialDim, Frame>& inv_metric)
     -> tnsr::Ijj<DataType, SpatialDim, Frame>;
 /// @}
+
+/// @{
+/*!
+ * \brief Compute \f$\Gamma_a\f$ from the generalized harmonic evolved
+ * variables.
```
Comment (geoffrey4444) [inline src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.hpp:47]: Do you have an equation + paper to cite for these equations?

File: src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp (hunk 1 above)
Comment (geoffrey4444) [reply src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Christoffel.cpp:76]: I'm also OK doing for loops for this one

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp [hunk 3]
```diff
@@ -201,6 +205,20 @@ void test_compute_extrinsic_curvature_and_deriv_metric(const T& used_for_size) {
...
+  const auto trace_christoffel = GeneralizedHarmonic::trace_christoffel(
+      spacetime_normal_one_form, spacetime_normal_vector,
+      inverse_spatial_metric, inverse_spacetime_metric, pi, phi);
+  tnsr::abb<T, Dim, Frame::Inertial> d4_spacetime_metric;
+  ::GeneralizedHarmonic::spacetime_derivative_of_spacetime_metric(
+      make_not_null(&d4_spacetime_metric), lapse, shift, pi, phi);
+  const auto expected_trace_christoffel =
+      trace_last_indices(gr::christoffel_first_kind(d4_spacetime_metric),
+                         inverse_spacetime_metric);
+  CHECK_ITERABLE_APPROX(trace_christoffel, expected_trace_christoffel);
```
Comment (geoffrey4444) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp:221]: (Optional) If I'm paranoid, I could imagine this code and the trace function being broken together...maybe add a python/einsum test as well to catch such cases?

**Review by geoffrey4444 — DISMISSED** [review]
Happy to merge as-is, but I have some optional comments for Nils to consider. I did not check that the LaTeX equations in the documentation are correct or that the implementation does implement those equations...I think the unit test is convincing, but if you want me to check by hand that I believe all the equations, I could do that, though it would take me some time.

**Comment by nilsdeppe** [comment]
Rebased, squashed. Added reference and python test. Thanks for the reviews :)

**Review by geoffrey4444 — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**